x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/InternalRealms.java

```diff
@@ -52,13 +53,18 @@
 public final class InternalRealms {
 
     /**
-     * The list of all <em>internal</em> realm types, excluding {@link ReservedRealm#TYPE}.
+     * The map of all <em>internal</em> realm types, excluding {@link ReservedRealm#TYPE}, to their relevant license feature
      */
-    private static final Set<String> XPACK_TYPES = Collections
-        .unmodifiableSet(Sets.newHashSet(NativeRealmSettings.TYPE, FileRealmSettings.TYPE, LdapRealmSettings.AD_TYPE,
-            LdapRealmSettings.LDAP_TYPE, PkiRealmSettings.TYPE, SamlRealmSettings.TYPE, KerberosRealmSettings.TYPE,
-            OpenIdConnectRealmSettings.TYPE));
-
+    private static final Map<String, XPackLicenseState.Feature> XPACK_TYPES = Map.ofEntries(
+        Map.entry(NativeRealmSettings.TYPE, XPackLicenseState.Feature.SECURITY),
+        Map.entry(FileRealmSettings.TYPE, XPackLicenseState.Feature.SECURITY),
+        Map.entry(LdapRealmSettings.AD_TYPE, XPackLicenseState.Feature.SECURITY_AD_REALM),
+        Map.entry(LdapRealmSettings.LDAP_TYPE, XPackLicenseState.Feature.SECURITY_LDAP_REALM),
+        Map.entry(PkiRealmSettings.TYPE, XPackLicenseState.Feature.SECURITY_PKI_REALM),
+        Map.entry(SamlRealmSettings.TYPE, XPackLicenseState.Feature.SECURITY_SAML_REALM),
+        Map.entry(OpenIdConnectRealmSettings.TYPE, XPackLicenseState.Feature.SECURITY_OIDC_REALM),
+        Map.entry(KerberosRealmSettings.TYPE, XPackLicenseState.Feature.SECURITY_KERBEROS_REALM)
+    );
```

Comment:
Other than `InternalRealms#getConfigurableRealmsTypes`, is there any other reason for this map to not include the `ReservedRealm`? In other places where this map is used, e.g. `isXPackRealm`, it feels better if `ReversedRealm` is in it. 

`getConfigurableRealmsTypes` seems to be the only exception. It is only used for tests as far as I can tell. Would it be better to remove `ReservedRealm` inside this method only instead of not having it in the Map?

Reply:
Probably. There's a lot that could be cleaned up here. I tried to keep this PR as self contained as possible and didn't want to change the meaning of `XPACK_TYPES` unnecessarily.